- Author: Yousef Al Zeer

* This task aims to analyze the impact of feature engineering on model performance.

# Imports

In [38]:
# Imports
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import make_pipeline

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
fpath = "/content/drive/MyDrive/Intermediate ML /Week 1 /Core/bikeshare_train - bikeshare_train.csv"
df = pd.read_csv(fpath)
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   datetime    10886 non-null  object 
 1   season      10886 non-null  int64  
 2   holiday     10886 non-null  int64  
 3   workingday  10886 non-null  int64  
 4   weather     10886 non-null  int64  
 5   temp        10886 non-null  float64
 6   atemp       10886 non-null  float64
 7   humidity    10886 non-null  int64  
 8   windspeed   10886 non-null  float64
 9   casual      10886 non-null  int64  
 10  registered  10886 non-null  int64  
 11  count       10886 non-null  int64  
dtypes: float64(3), int64(8), object(1)
memory usage: 1020.7+ KB


,datetime,season,holiday,workingday,weather,temp,atemp,humidity,windspeed,casual,registered,count
0,2011-01-01 0:00:00,1,0,0,1,9.84,14.395,81,0.0,3,13,16
1,2011-01-01 1:00:00,1,0,0,1,9.02,13.635,80,0.0,8,32,40
2,2011-01-01 2:00:00,1,0,0,1,9.02,13.635,80,0.0,5,27,32
3,2011-01-01 3:00:00,1,0,0,1,9.84,14.395,75,0.0,3,10,13
4,2011-01-01 4:00:00,1,0,0,1,9.84,14.395,75,0.0,0,1,1


In [4]:
# drop  'casual' and 'registered' columns due to redundancy
df.drop(columns=['casual' , 'registered'] , inplace=True)

In [5]:
df_original = df.copy()

# Datetime Transformation

In [6]:
df['datetime'] = pd.to_datetime(df['datetime'])
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   datetime    10886 non-null  datetime64[ns]
 1   season      10886 non-null  int64         
 2   holiday     10886 non-null  int64         
 3   workingday  10886 non-null  int64         
 4   weather     10886 non-null  int64         
 5   temp        10886 non-null  float64       
 6   atemp       10886 non-null  float64       
 7   humidity    10886 non-null  int64         
 8   windspeed   10886 non-null  float64       
 9   count       10886 non-null  int64         
dtypes: datetime64[ns](1), float64(3), int64(6)
memory usage: 850.6 KB


In [7]:
df['month'] = df['datetime'].dt.month_name()
df['day of week'] = df['datetime'].dt.day_name()
df['hour of day'] = df['datetime'].dt.hour
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   datetime     10886 non-null  datetime64[ns]
 1   season       10886 non-null  int64         
 2   holiday      10886 non-null  int64         
 3   workingday   10886 non-null  int64         
 4   weather      10886 non-null  int64         
 5   temp         10886 non-null  float64       
 6   atemp        10886 non-null  float64       
 7   humidity     10886 non-null  int64         
 8   windspeed    10886 non-null  float64       
 9   count        10886 non-null  int64         
 10  month        10886 non-null  object        
 11  day of week  10886 non-null  object        
 12  hour of day  10886 non-null  int32         
dtypes: datetime64[ns](1), float64(3), int32(1), int64(6), object(2)
memory usage: 1.0+ MB


,datetime,season,holiday,workingday,weather,temp,atemp,humidity,windspeed,count,month,day of week,hour of day
0,2011-01-01 00:00:00,1,0,0,1,9.84,14.395,81,0.0,16,January,Saturday,0
1,2011-01-01 01:00:00,1,0,0,1,9.02,13.635,80,0.0,40,January,Saturday,1
2,2011-01-01 02:00:00,1,0,0,1,9.02,13.635,80,0.0,32,January,Saturday,2
3,2011-01-01 03:00:00,1,0,0,1,9.84,14.395,75,0.0,13,January,Saturday,3
4,2011-01-01 04:00:00,1,0,0,1,9.84,14.395,75,0.0,1,January,Saturday,4


In [8]:
df['hour of day'] = df['hour of day'].astype('object')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   datetime     10886 non-null  datetime64[ns]
 1   season       10886 non-null  int64         
 2   holiday      10886 non-null  int64         
 3   workingday   10886 non-null  int64         
 4   weather      10886 non-null  int64         
 5   temp         10886 non-null  float64       
 6   atemp        10886 non-null  float64       
 7   humidity     10886 non-null  int64         
 8   windspeed    10886 non-null  float64       
 9   count        10886 non-null  int64         
 10  month        10886 non-null  object        
 11  day of week  10886 non-null  object        
 12  hour of day  10886 non-null  object        
dtypes: datetime64[ns](1), float64(3), int64(6), object(3)
memory usage: 1.1+ MB


In [9]:
# Drop the 'datetime' and 'season' columns
df.drop(columns=['datetime' , 'season'] , inplace=True)

# Temp Transformation

$$
F = (C \times 1.8) + 32
$$

In [10]:
df['temp'] = df['temp'].apply(lambda temp: (temp * 1.8) + 32)
df['atemp'] = df['atemp'].apply(lambda temp: (temp * 1.8) + 32)

In [19]:
df['temp_variance'] = df['temp'] - df['atemp']
df[['temp' , 'atemp' , 'temp_variance']].head()

,temp,atemp,temp_variance
0,49.712,57.911,-8.199
1,48.236,56.543,-8.307
2,48.236,56.543,-8.307
3,49.712,57.911,-8.199
4,49.712,57.911,-8.199


In [17]:
df[['temp', 'atemp']].describe()

,temp,atemp
count,10886.000000,10886.000000
mean,68.415548,74.579151
std,14.024862,15.254281
min,33.476000,33.368000
25%,57.092000,61.997000
50%,68.900000,75.632000
75%,79.232000,87.908000
max,105.800000,113.819000


In [20]:
df.drop(columns=['atemp'] , inplace=True)

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   holiday        10886 non-null  int64  
 1   workingday     10886 non-null  int64  
 2   weather        10886 non-null  int64  
 3   temp           10886 non-null  float64
 4   humidity       10886 non-null  int64  
 5   windspeed      10886 non-null  float64
 6   count          10886 non-null  int64  
 7   month          10886 non-null  object 
 8   day of week    10886 non-null  object 
 9   hour of day    10886 non-null  object 
 10  temp_variance  10886 non-null  float64
dtypes: float64(3), int64(5), object(3)
memory usage: 935.6+ KB


# Modeling

In [33]:
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error, root_mean_squared_error
def regression_metrics(y_true, y_pred, label='', verbose = True, output_dict=False):
  # Get metrics
  mae = mean_absolute_error(y_true, y_pred)
  mse = mean_squared_error(y_true, y_pred)
  rmse = root_mean_squared_error(y_true, y_pred)
  r_squared = r2_score(y_true, y_pred)
  if verbose == True:
    # Print Result with Label and Header
    header = "-"*60
    print(header, f"Regression Metrics: {label}", header, sep='\n')
    print(f"- MAE = {mae:,.3f}")
    print(f"- MSE = {mse:,.3f}")
    print(f"- RMSE = {rmse:,.3f}")
    print(f"- R^2 = {r_squared:,.3f}")
  if output_dict == True:
      metrics = {'Label':label, 'MAE':mae,
                 'MSE':mse, 'RMSE':rmse, 'R^2':r_squared}
      return metrics



def evaluate_regression(reg, X_train, y_train, X_test, y_test, verbose = True,
                        output_frame=False):
  # Get predictions for training data
  y_train_pred = reg.predict(X_train)

  # Call the helper function to obtain regression metrics for training data
  results_train = regression_metrics(y_train, y_train_pred, verbose = verbose,
                                     output_dict=output_frame,
                                     label='Training Data')
  print()
  # Get predictions for test data
  y_test_pred = reg.predict(X_test)
  # Call the helper function to obtain regression metrics for test data
  results_test = regression_metrics(y_test, y_test_pred, verbose = verbose,
                                  output_dict=output_frame,
                                    label='Test Data' )

  # Store results in a dataframe if ouput_frame is True
  if output_frame:
    results_df = pd.DataFrame([results_train,results_test])
    # Set index.name to none to get a cleaner looking result
    results_df.index.name=None
    # Return the dataframe
    return results_df.round(3)

## Original featrues

In [28]:
df_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10886 entries, 0 to 10885
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   datetime    10886 non-null  object 
 1   season      10886 non-null  int64  
 2   holiday     10886 non-null  int64  
 3   workingday  10886 non-null  int64  
 4   weather     10886 non-null  int64  
 5   temp        10886 non-null  float64
 6   atemp       10886 non-null  float64
 7   humidity    10886 non-null  int64  
 8   windspeed   10886 non-null  float64
 9   count       10886 non-null  int64  
dtypes: float64(3), int64(6), object(1)
memory usage: 850.6+ KB


In [29]:
y = df_original['count']
X = df_original.drop(columns=['count' , 'datetime'])
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 42)

In [30]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8164 entries, 2930 to 7270
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   season      8164 non-null   int64  
 1   holiday     8164 non-null   int64  
 2   workingday  8164 non-null   int64  
 3   weather     8164 non-null   int64  
 4   temp        8164 non-null   float64
 5   atemp       8164 non-null   float64
 6   humidity    8164 non-null   int64  
 7   windspeed   8164 non-null   float64
dtypes: float64(3), int64(5)
memory usage: 574.0 KB


In [31]:
from sklearn.ensemble import RandomForestRegressor

rf_model = RandomForestRegressor(random_state=42)
rf_model.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

In [34]:
evaluate_regression(rf_model, X_train, y_train, X_test, y_test)

------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 47.863
- MSE = 5,084.585
- RMSE = 71.306
- R^2 = 0.845

------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 107.423
- MSE = 23,153.867
- RMSE = 152.164
- R^2 = 0.294


## Feature Engineering

In [35]:
y = df['count']
X = df.drop(columns=['count'])
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 42)

In [36]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8164 entries, 2930 to 7270
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   holiday        8164 non-null   int64  
 1   workingday     8164 non-null   int64  
 2   weather        8164 non-null   int64  
 3   temp           8164 non-null   float64
 4   humidity       8164 non-null   int64  
 5   windspeed      8164 non-null   float64
 6   month          8164 non-null   object 
 7   day of week    8164 non-null   object 
 8   hour of day    8164 non-null   object 
 9   temp_variance  8164 non-null   float64
dtypes: float64(3), int64(4), object(3)
memory usage: 701.6+ KB


### Encoding

In [40]:
cat_cols = ['month', 'day of week', 'hour of day']
ohe_encode = OneHotEncoder(handle_unknown='ignore', sparse_output=False)
ohe_pipe = make_pipeline(ohe_encode)
ohe_tuple = ("Categorical" , ohe_pipe , cat_cols)
ohe_tuple

('Categorical',
 Pipeline(steps=[('onehotencoder',
                  OneHotEncoder(handle_unknown='ignore', sparse_output=False))]),
 ['month', 'day of week', 'hour of day'])

In [41]:
col_transformer = ColumnTransformer([ohe_tuple ], verbose_feature_names_out=False)

In [42]:
rf_model = RandomForestRegressor(random_state=42)
rf_model_pipe = make_pipeline(col_transformer , rf_model)
rf_model_pipe.fit(X_train , y_train)

Pipeline(steps=[('columntransformer',
                 ColumnTransformer(transformers=[('Categorical',
                                                  Pipeline(steps=[('onehotencoder',
                                                                   OneHotEncoder(handle_unknown='ignore',
                                                                                 sparse_output=False))]),
                                                  ['month', 'day of week',
                                                   'hour of day'])],
                                   verbose_feature_names_out=False)),
                ('randomforestregressor',
                 RandomForestRegressor(random_state=42))])

In [43]:
evaluate_regression(rf_model_pipe, X_train, y_train, X_test, y_test)

------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 51.529
- MSE = 6,185.600
- RMSE = 78.649
- R^2 = 0.811

------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 69.504
- MSE = 11,121.569
- RMSE = 105.459
- R^2 = 0.661



| Model | Dataset | Train R² | Test R² | MAE (Test) | RMSE (Test) |
|------|----------|---------|--------|------------|-------------|
| Random Forest | Original Features | 0.845 | 0.294 | 107.423 | 152.164 |
| Random Forest | Feature Engineered | 0.811 | 0.661 | 69.504 | 105.459 |



* Feature engineering significantly improved model performance overall metrics, reducing overfitting and improving accuracy